open-craft/edx-platform
# Adaptive Content Block: Send "Unit viewed" event, selectively display review questions

This PR extends the new Adaptive Content Block in two ways:

* It adds logic for notifying Domoscio that the parent unit of an ACB has been accessed by a learner.
* It changes the way child blocks are selected for display: Instead of selecting `max_count` children randomly, ACBs now query Domoscio for information about pending reviews to determine which children to show (if any).

The PR also adds a new course-wide setting called "Adaptive Learning Configuration" which can be used to specify necessary information for communicating with the Domoscio API.

cf. [OC-2150](https://tasks.opencraft.com/browse/OC-2150), [OC-2151](https://tasks.opencraft.com/browse/OC-2151)

**Implementation**

When sending a "Unit viewed" event, ACB assumes that the corresponding unit exists in Domoscio's database (in the form of a "knowledge node" whose `uid` is set to the `block_id` of the unit). This is a safe assumption to make because Domoscio staff will take care of creating entities representing relevant elements of the course structure (including units and review questions) prior to the start date of the course that will be using the new adaptive learning features.

In addition to knowledge nodes representing units, there are other entities that must exist in the Domoscio database to be able to send an event of the type that Domoscio expects: Events target objects that link students to specific knowledge nodes. This means that prior to sending a "Unit viewed" event, the Domoscio database must contain both a "student" entity corresponding to the learner viewing the unit in the LMS, and an object that links the student to the unit (knowledge node). In Domoscio terminology, an object linking a student to a knowledge node is called "knowledge node student". (The plural is "knowledge node students".) Since we can't assume that students and objects that link them to units will be created by Domoscio staff prior to the start date of the course, ACBs will create student and "knowledge node student" objects via the Domoscio API prior to sending "Unit viewed" events (if necessary).

As of this PR, ACBs will also create "knowledge node student" objects that link students to review questions when a learner accesses a unit containing an ACB. This is not a prerequisite for sending "Unit viewed" events (or for obtaining a list of review questions to display from Domoscio). However, when a learner submits an answer to a review question, we will need to notify Domoscio about the result (OC-2152). At that point, links between students and review questions need to exist so we can target them with appropriate `EventResult` events.

**Test instructions**

1. Follow instructions on the [epic](https://tasks.opencraft.com/browse/OC-2127) to set up a FUN box.

2. Switch to this branch (`adaptive-content-block`) and install the new block via

    ```sh
    pip install -r requirements/edx/local.txt
    ```

3. Create a new content library and add a few problems to it.

4. Create a new course and add `"adaptive_library_content"` to advanced modules.

5. Add an "Adaptive Content Block" that references the library created previously to a unit (optionally modifying the number of components to display to each student), and publish the unit. Do **not** access the unit in the LMS yet.

6. To be able to work with the Domoscio API, you'll need to specify a few course-level settings: In Studio, navigate to the Advanced Settings page for the course ("Paramètres" > "Paramètres avancés") and replace the default value for the "Adaptive Learning Configuration" setting with the following:

    ```json
    "url": "<see comment on internal ticket>",
    "instance_id": <see comment on internal ticket>,
    "access_token": "<see comment on internal ticket>",
    "api_version": "v1"
    ```

7. In a venv that has the `requests` library installed, create a knowledge node that corresponds to the unit you created earlier (`block_id` needs to be a 32-character hash), and review the results:

    ```python
    import requests, json
    response = requests.post("<url>/v1/instances/<instance_id>/knowledge_nodes", headers={"Authorization": "Token token=<access_token>"}, data={"knowledge_graph_id": <see comment on internal ticket>, "name": "Unit node", "uid": "<block_id of unit>"})
    json.loads(response.content)
    ```

    Then create a knowledge node for each of the problems that belong to the library you created earlier. Note that problems belonging to a content library are assigned unique block IDs for each Adaptive Content Block that references them; you'll need to use these context-specific IDs when creating corresponding knowledge nodes. Block IDs for problems are 20-character hashes.

8. Review the list of students that exist on the Domoscio test instance:

   ```python
   response = requests.get("<url>/v1/instances/<instance_id>/students", headers={"Authorization": "Token token=<access_token>"})
   json.loads(response.content)
   ```

9. Review the list of links between students and units ("knowledge node student" objects) on the Domoscio test instance:

    ```python
    response = requests.get("<url>/v1/instances/<instance_id>/knowledge_node_students", headers={"Authorization": "Token token=<access_token>"})
    json.loads(response.content)    
    ```

10. Access the unit that contains the ACB in the LMS. Observe that it does not display any child blocks.

11. Review the list of students and the list of "knowledge node student" objects as described above. Observe that the former contains a new student and the latter contains new "knowledge node student" objects, one for each knowledge node you created earlier. The `knowledge_node_id`/`knowledge_node_uid` and `student_id`/`student_uid` fields of the new "knowledge node student" objects should reference the knowledge nodes created earlier, and the student that was created when you accessed the unit for the first time.

12. Review the list of events that have been recorded for the Domoscio test instance, making sure that it contains an entry that references the "knowledge node student" object linking the unit you created earlier to the student that you used to access the unit in the LMS:

    ```python
    response = requests.get("<url>/v1/instances/<instance_id>/events", headers={"Authorization": "Token token=<access_token>"})
    json.loads(response.content)
    ```

**Reviewers**

- [ ] @e-kolpakov / @pomegranited

**Notes**


- There is no way to force Domoscio to schedule a review, so it's not possible to manually test the logic for selectively displaying review questions (except the case where no pending reviews are available).
